File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/layernorm2d_emit.c
/* Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
 * SPDX-License-Identifier: MIT
 *
 * tests/parity/layernorm2d_emit.c -- C-side emitter for the LayerNorm2D parity
 * STRESS harness. Selects one config by argv[1] (the config index), builds
 * rocke_layernorm2d_spec_t identically to the Python emitter layernorm2d_emit.py,
 * lowers via rocke_layernorm2d_lower_to_llvm (arch gfx950, flavor AUTO) and
 * prints the .ll to stdout so the two outputs can be byte-compared.
 *
 * The config table MUST stay in lockstep with CONFIGS in layernorm2d_emit.py.
 *
 * Optional argv[2] selects the output mode:
 *   "ll"     (default) - lower to LLVM and print
 *   "ir"               - print ck.dsl.ir/v1 serialization
 *   "verify"           - run verifier; print each diagnostic on its own line
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "rocke/instance_layernorm2d.h"
#include "rocke/ir.h"
#include "rocke/ir_serialize.h"
#include "rocke/lower_llvm.h"
#include "rocke/verify.h"

typedef struct
{
    int n_per_block;
    int block_size;
    int vec;
    const char* dtype;
    bool save_mean_invstd;
} ln_cfg_t;

/* MUST match CONFIGS in layernorm2d_emit.py (index for index). */
static const ln_cfg_t CONFIGS[] = {
    /* 0..5  original sampled */
    {4096, 256, 4, "f16", false},
    {4096, 256, 8, "f16", false},
    {4096, 256, 4, "bf16", false},
    {2048, 128, 4, "f16", true},
    {8192, 256, 8, "f16", false},
    {1024, 256, 2, "bf16", true},
    /* 6..9 tiny */
    {128, 64, 2, "f16", false},
    {256, 64, 4, "f16", true},
    {512, 64, 8, "bf16", false},
    {128, 64, 2, "bf16", true},
    /* 10..13 block sizes */
    {4096, 512, 4, "f16", false},
    {8192, 1024, 8, "f16", false},
    {2048, 1024, 2, "bf16", true},
    {1024, 128, 8, "f16", false},
    /* 14..15 fp16 alias */
    {4096, 256, 4, "fp16", false},
    {2048, 128, 2, "fp16", true},
    /* 16..21 odd multipliers */
    {1536, 256, 2, "f16", false},
    {3072, 256, 4, "bf16", false},
    {5120, 256, 4, "f16", true},
    {1792, 128, 2, "bf16", false},
    {2816, 128, 2, "f16", false},
    {6656, 256, 2, "bf16", true},
    /* 22..27 two-pass */
    {16384, 256, 8, "f16", false},
    {33280, 256, 2, "f16", false},
    {32768, 256, 8, "bf16", false},
    {65536, 256, 8, "f16", true},
    {131072, 512, 8, "bf16", false},
    {34816, 256, 2, "bf16", true},
    /* 28..29 very large single/two pass at 1024 block */
    {65536, 1024, 8, "f16", false},
    {133120, 1024, 2, "bf16", false},
    /* 30..32 vec sweep */
    {2048, 256, 2, "f16", false},
    {4096, 256, 4, "f16", true},
    {8192, 256, 8, "bf16", true},
    /* 33..34 block 512 two-pass */
    {66560, 512, 2, "f16", false},
    {133120, 512, 4, "bf16", true},
};

#define NCFG ((int)(sizeof(CONFIGS) / sizeof(CONFIGS[0])))

static int make_spec(int idx, rocke_layernorm2d_spec_t* spec)
{
    if(idx < 0 || idx >= NCFG)
        return -1;
    *spec = rocke_layernorm2d_spec_default();
    spec->n_per_block = CONFIGS[idx].n_per_block;
    spec->block_size = CONFIGS[idx].block_size;
    spec->vec = CONFIGS[idx].vec;
    spec->dtype = CONFIGS[idx].dtype;
    spec->save_mean_invstd = CONFIGS[idx].save_mean_invstd;
    return 0;
}

int main(int argc, char** argv)
{
    if(argc < 2)
    {
        fprintf(stderr, "usage: %s <config_index 0..%d> [ll|ir|verify]\n", argv[0], NCFG - 1);
        return 2;
    }
    if(strcmp(argv[1], "--count") == 0)
    {
        printf("%d\n", NCFG);
        return 0;
    }
    int idx = atoi(argv[1]);
    const char* mode = (argc > 2) ? argv[2] : "ll";

    if(strcmp(mode, "ll") != 0 && strcmp(mode, "ir") != 0 && strcmp(mode, "verify") != 0)
    {
        fprintf(stderr, "unknown mode %s\n", mode);
        return 2;
    }

    rocke_layernorm2d_spec_t spec;
    if(make_spec(idx, &spec) != 0)
    {
        fprintf(stderr, "unknown config index %d\n", idx);
        return 2;
    }

    rocke_ir_builder_t b;
    rocke_kernel_def_t* kernel = rocke_build_layernorm2d_new(&b, &spec);
    if(kernel == NULL)
    {
        const char* m = rocke_ir_builder_error(&b);
        fprintf(stderr, "build failed: %s\n", m ? m : "(no message)");
        rocke_ir_builder_free(&b);
        return 1;
    }

    if(strcmp(mode, "ll") == 0)
    {
        char* llvm_text = NULL;
        rocke_status_t st
            = rocke_lower_kernel_to_llvm(kernel, ROCKE_LLVM_FLAVOR_AUTO, "gfx950", &llvm_text);
        if(st != ROCKE_OK || !llvm_text)
        {
            fprintf(stderr, "lower failed: status=%d\n", (int)st);
            rocke_ir_builder_free(&b);
            return 1;
        }
        fputs(llvm_text, stdout);
        free(llvm_text);
    }
    else if(strcmp(mode, "ir") == 0)
    {
        char* text = NULL;
        rocke_status_t st = rocke_ir_serialize(kernel, &text);
        if(st != ROCKE_OK || !text)
        {
            fprintf(stderr, "serialize failed: status=%d\n", (int)st);
            rocke_ir_builder_free(&b);
            return 1;
        }
        fputs(text, stdout);
        free(text);
    }
    else
    { /* verify */
        rocke_diag_t* d = NULL;
        size_t n = 0;
        rocke_verify(kernel, &d, &n);
        for(size_t i = 0; i < n; i++)
        {
            char* s = rocke_diag_to_string(&d[i]);
            if(s)
            {
                puts(s);
                free(s);
            }
        }
        rocke_diags_free(d, n);
    }

    rocke_ir_builder_free(&b);
    return 0;
}

File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/layernorm2d_emit.py
#!/usr/bin/env python3
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
#
# tests/parity/layernorm2d_emit.py -- Python reference emitter for the
# LayerNorm2D parity STRESS harness. Selects one of N configs by argv[1]
# (the config index) and prints _native_lower(arch='gfx950') to stdout
# so it can be byte-compared with the C emitter layernorm2d_emit.c.
import sys

from rocke.instances.common.layernorm2d import (
    LayerNorm2DSpec,
    build_layernorm2d,
)

try:
    from rocke.core.lower_llvm import _lower_kernel_to_llvm_python as _native_lower
except ImportError:  # pragma: no cover - older reference tree
    from rocke import lower_kernel_to_llvm as _native_lower
from rocke.core.ir_serialize import serialize
from rocke.core.verify import verify


# (n_per_block, block_size, vec, dtype, save_mean_invstd)
CONFIGS = [
    # --- original 6 sampled configs (regression anchor) ---
    (4096, 256, 4, "f16", False),  # 0
    (4096, 256, 8, "f16", False),  # 1
    (4096, 256, 4, "bf16", False),  # 2
    (2048, 128, 4, "f16", True),  # 3
    (8192, 256, 8, "f16", False),  # 4
    (1024, 256, 2, "bf16", True),  # 5
    # --- tiny / minimal shapes (edge: smallest valid N) ---
    (128, 64, 2, "f16", False),  # 6  N = block*vec, elems=2
    (256, 64, 4, "f16", True),  # 7
    (512, 64, 8, "bf16", False),  # 8
    (128, 64, 2, "bf16", True),  # 9
    # --- all block sizes ---
    (4096, 512, 4, "f16", False),  # 10
    (8192, 1024, 8, "f16", False),  # 11  block_size at the 1024 cap
    (2048, 1024, 2, "bf16", True),  # 12
    (1024, 128, 8, "f16", False),  # 13
    # --- fp16 alias (validate_io allows "fp16"; Python Literal is f16/bf16) ---
    (4096, 256, 4, "fp16", False),  # 14  ADVERSARIAL: aliasing
    (2048, 128, 2, "fp16", True),  # 15
    # --- odd / non-power-of-2 multipliers (edge, still divisible) ---
    (1536, 256, 2, "f16", False),  # 16  1536 = 256*2*3
    (3072, 256, 4, "bf16", False),  # 17  3072 = 256*4*3
    (5120, 256, 4, "f16", True),  # 18  5120 = 256*4*5 (prime factor 5)
    (1792, 128, 2, "bf16", False),  # 19  1792 = 128*2*7 (prime factor 7)
    (2816, 128, 2, "f16", False),  # 20  2816 = 128*2*11 (prime factor 11)
    (6656, 256, 2, "bf16", True),  # 21  6656 = 256*2*13 (prime factor 13)
    # --- two-pass path: elems_per_thread > 64 (re-stream X in pass 2) ---
    (16384, 256, 8, "f16", False),  # 22  elems=64 (boundary, single-pass)
    (33280, 256, 2, "f16", False),  # 23  elems=130 -> TWO-PASS (256*2*65)
    (32768, 256, 8, "bf16", False),  # 24  elems=128 -> TWO-PASS
    (65536, 256, 8, "f16", True),  # 25  elems=256 -> TWO-PASS, save mv
    (131072, 512, 8, "bf16", False),  # 26  elems=256 -> TWO-PASS, big block
    (34816, 256, 2, "bf16", True),  # 27  elems=136 two-pass + save mv bf16 (256*2*68)
    # --- very large N single-pass at high block_size ---
    (65536, 1024, 8, "f16", False),  # 28  elems=64 boundary at 1024 block
    (133120, 1024, 2, "bf16", False),  # 29 elems=130 two-pass at 1024 block (1024*2*65)
    # --- vec sweep at fixed shape ---
    (2048, 256, 2, "f16", False),  # 30
    (4096, 256, 4, "f16", True),  # 31  same as 0 but save mv
    (8192, 256, 8, "bf16", True),  # 32
    # --- block 512 + two-pass ---
    (66560, 512, 2, "f16", False),  # 33  elems=130 two-pass, block 512 (512*2*65)
    (133120, 512, 4, "bf16", True),  # 34  elems=260 two-pass + save mv (512*4*65)
]


def _spec(idx: int) -> LayerNorm2DSpec:
    if idx < 0 or idx >= len(CONFIGS):
        raise SystemExit(f"unknown config index {idx}")
    n, bs, v, dt, smv = CONFIGS[idx]
    return LayerNorm2DSpec(
        n_per_block=n, block_size=bs, vec=v, dtype=dt, save_mean_invstd=smv
    )


def main() -> int:
    if len(sys.argv) < 2:
        sys.stderr.write(
            f"usage: layernorm2d_emit.py <config_index 0..{len(CONFIGS) - 1}> [ll|ir|verify]\n"
        )
        return 2
    if sys.argv[1] == "--count":
        sys.stdout.write(str(len(CONFIGS)) + "\n")
        return 0
    idx = int(sys.argv[1])
    mode = sys.argv[2] if len(sys.argv) > 2 else "ll"
    if mode not in ("ll", "ir", "verify"):
        sys.stderr.write(f"unknown mode {mode}\n")
        return 2
    spec = _spec(idx)
    kernel = build_layernorm2d(spec)
    if mode == "ll":
        text = _native_lower(kernel, arch="gfx950")
        sys.stdout.write(text)
    elif mode == "ir":
        sys.stdout.write(serialize(kernel))
    else:  # verify
        sys.stdout.write("".join(str(d) + "\n" for d in verify(kernel)))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
